src/mmr.rs

```diff
@@ -271,37 +270,22 @@ impl<T: Clone + PartialEq, M: Merge<Item = T>, S: MMRStoreReadOps<T>> MMR<T, M,
                 }
             };
 
-            let queue_front_pos = queue.front().map(|(pos, _)| pos);
-            if Some(&sib_pos) == queue_front_pos {
+            if Some(&sib_pos) == queue.front().map(|(_, pos)| pos) {
                 // drop sibling
                 queue.pop_front();
-            } else if queue_front_pos.is_none()
-                || !is_descendant_pos(
-                    sib_pos,
-                    *queue_front_pos.expect("checked queue_front_pos != None"),
-                )
-            // only push a sibling into the proof if either of these cases is satisfied:
-            // 1. the queue is empty
-            // 2. the next item in the queue is not the sibling or a child of it
-            {
+            } else {
                 let sibling = (
                     sib_pos,
                     self.batch
                         .get_elem(sib_pos.clone())?
                         .ok_or(Error::InconsistentStore)?,
                 );
 
-                // only push sibling if it's not already a proof item or to be proven,
-                // which can be the case if both a child and its parent are to be proven
-                if height == 0
-                    || !(proof.contains(&sibling)) && pos_list.binary_search(&sib_pos).is_err()
-                {
-                    proof.push(sibling);
-                }
+                proof.push(sibling);
```

Comment:
The `proof.contains(&sibling)` check is no longer required? I see that this `gen_node_proof_for_peak` is called from a loop, so is that possible that we've pushed the same `sibling` in one of previous iterations?

Reply:
It's no longer required, because now we check for duplicates when inserting in the queue:

```
    fn insert_sorted(&mut self, value: T) {
        match self.binary_search(&value) {
            Ok(_pos) => {
                // element already in vector @ `pos`
            }
            Err(pos) => self.insert(pos, value),
        }
    }
```

Reply:
But the `queue` is only actual for a single `gen_node_proof_for_peak` call. I.e.:
```rust
fn gen_node_proof_for_peak() {
  let queue = ...;
  queue.insert_sorted();
}
```
And the same `proof` argument can be passed to multiple `gen_node_proof_for_peak` calls. I.e.:
```rust
let mut proof = Vec::new();
gen_node_proof_for_peak(&mut proof);
gen_node_proof_for_peak(&mut proof);
```
So even though you check duplicates when inserting to the queue, it has nothing common with the values from the `proof `, right? Or am I missing something?

Reply:
Yes, but `gen_node_proof_for_peak()` generates a proof only for the elements that are under that peak. The proofs for elements that are under different peaks shouldn't have elements in common.

Reply:
Ok, ty for explaining :)